### PR TITLE
Global thresholds

### PR DESCRIPTION
**Pull request by jewbetcha**
- Adds a form on the checks list to set threshold settings (for list view only)
- Adds uptime gauge to checks list
- Viz view reflects thresholds as well
- Works in tandem with new api routes (https://github.com/grafana/synthetic-monitoring-api/pull/327)

### CODE REVIEW

File: src/components/CheckList.tsx [hunk 1]
```diff
@@ -648,6 +661,12 @@ export const CheckList = ({ instance, onAddNewClick, checks, onCheckUpdate }: Pr
           )}
         </div>
       )}
+      <ThresholdGlobalSettings
+        onDismiss={() => setShowThresholdModal(false)}
```
Comment (rdubrock) [inline src/components/CheckList.tsx:665]: question: does onDismiss get called onSuccess? Or I guess the more relevant question is do we want the modal to close on successful select? I could go either way, just thinking aloud

File: src/components/ChecksVisualization/ChecksVisualization.tsx [hunk 2]
```diff
@@ -132,13 +132,13 @@ export function ChecksVisualization({ checks, showIcons }: Props) {
         <div>{hoveredCheck?.target}</div>
         <div className={styles.successGaugeContainer}>
           <SuccessRateGauge
```
Comment (rdubrock) [inline src/components/ChecksVisualization/ChecksVisualization.tsx:134]: [random thinking for a future pr] we could put the other stats in this tooltip too. Maybe we could let folks choose which stat informs the color of the visualization too?

File: src/components/SuccessRateContextProvider.tsx [hunk 3]
```diff
@@ -77,17 +100,16 @@ const parseProbeResults = (probes: Probe[] | undefined, data: any) => {
     return acc;
   }, {});
 
+  // TODO: Re-type latency check data
```
Comment (rdubrock) [inline src/components/SuccessRateContextProvider.tsx:103]: Is this still TODO? Fine if so and you want to do a follow up PR

File: src/components/Thresholds/ThresholdFormSection.tsx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+const getSectionStyles = () => (theme: GrafanaTheme2) => ({
+  estimate: css`
+    margin-bottom: 0px;
+    font-style: italic;
+    align-self: center;
+    font-size: 0.7rem;
+  `,
+  disabled: css`
+    // This is a hack, can't seem to get the disabled prop on Input to work
```
Comment (rdubrock) [inline src/components/Thresholds/ThresholdFormSection.tsx:50]: I think you can get the disabled state to work by wrapping the `<Input />` in a `<Field  disabled={disabled} />` component. I don't _think_ that `Field` has to exist in a form, so you should be able to do it stand-alone. See [here](https://github.com/grafana/synthetic-monitoring-app/blob/efde3227caa8784f7e3677de1ef7a5eec57b3140/src/components/CheckTarget.tsx#L67)

File: src/components/Thresholds/ThresholdFormSection.tsx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+      const newThresholds = {
+        ...thresholds,
+        [key]: newValue,
+      };
+      setThresholds(newThresholds);
+    },
+    [thresholds, setThresholds]
+  );
+
+  const After = <div style={{ display: 'flex', alignItems: 'center', padding: '5px' }}>{unit}</div>;
```
Comment (rdubrock) [inline src/components/Thresholds/ThresholdFormSection.tsx:106]: [nit] consider adding these styles to your `getStyles` func and passing as a classname

File: src/components/Thresholds/ThresholdGlobalSettings.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+  // Set thresholds in form when they are updated in context
+  useEffect(() => {
+    setUptimeThresholds(thresholds.uptime);
+    setReachabilityThresholds(thresholds.reachability);
+    setLatencyThresholds(thresholds.latency);
+  }, [thresholds]);
+
+  return (
+    <Modal title="Threshold Settings" isOpen={isOpen} onDismiss={onDismiss}>
```
Comment (rdubrock) [inline src/components/Thresholds/ThresholdGlobalSettings.tsx:84]: I think I would prefer if this were a form so we could delegate all the state management shenanigans. Not a strong feeling, and it's up to you. Definitely don't need to do it in this PR.

File: src/components/Thresholds/ThresholdGlobalSettings.test.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  // Latency
+  expect(upperLimitInputs[2]).toHaveValue(5);
+  expect(lowerLimitInputs[2]).toHaveValue(10);
+});
+
+test('submits the form', async () => {
+  const instance = renderThresholdSettingsForm();
+  const saveButton = await screen.getByTestId('threshold-save');
+  userEvent.click(saveButton);
+  expect(instance.api?.updateTenantSettings).toHaveBeenCalled();
```
Comment (rdubrock) [inline src/components/Thresholds/ThresholdGlobalSettings.test.tsx:59]: Maybe do a `toHaveBeenCalledWith({...})` so you can assert that the values are getting processed correctly.

File: src/contexts/SuccessRateContext.ts [hunk 8]
```diff
@@ -30,35 +38,54 @@ export interface SuccessRates extends SuccessRatesByType {
...
+    uptimeValue: 0,
+    uptimeDisplayValue: 'N/A',
   },
 };
 
-const updateSuccessRate = (type: SuccessRateTypes, id: number, successRate: number | undefined) => {
-  const thresholdColor = getSuccessRateThresholdColor(successRate);
+export const defaultThresholds: ThresholdSettings = {
+  uptime: {
+    upperLimit: 90,
```
Comment (rdubrock) [inline src/contexts/SuccessRateContext.ts:60]: Are these the right defaults? I think the current defaults are 99.5, 99, and 75? Not sure. The latency defaults also seem a little small (if they are in `ms`)

File: src/utils.ts [hunk 9]
```diff
@@ -276,28 +277,26 @@ export const fromBase64 = (value: string) => {
...
-  if (value >= 99) {
-    return config.theme2.colors.warning.main;
-  }
-  return config.theme2.colors.error.main;
+export const getSuccessRateThresholdColor = (
+  thresholds: ThresholdSettings,
+  key: 'reachability' | 'uptime' | 'latency',
+  compareValue: number
+) => {
+  return compareValue < thresholds[key].lowerLimit
```
Comment (rdubrock) [inline src/utils.ts:285]: Do you mind busting this out into `if` statements? Nested ternaries make my brain hurt

File: src/utils.ts [hunk 10]
```diff
@@ -276,28 +277,26 @@ export const fromBase64 = (value: string) => {
...
-  if (value >= 99) {
-    return 'exclamation-triangle';
-  }
-  return 'times-square' as IconName;
+export const getSuccessRateIcon = (
+  thresholds: ThresholdSettings,
+  key: 'reachability' | 'uptime' | 'latency',
+  compareValue: number
+): IconName => {
+  return compareValue < thresholds[key].lowerLimit
```
Comment (rdubrock) [inline src/utils.ts:297]: Nested ternaries again. Really it's fine, just a personal preference thing.

**Review by rdubrock — APPROVED** [review]
Awesome work! I left some commentary, I don't think any of it is blocking necessarily, so I'm going ahead and approving.

File: src/components/CheckList.tsx (hunk 1 above)
Comment (jewbetcha) [reply src/components/CheckList.tsx:665]: Yep, upon successful update, `onSuccess` and `onDismiss` are called in succession.

https://github.com/grafana/synthetic-monitoring-app/pull/338/files#diff-e435217755667714e83a32a6ccbf2072b56705993b58c69d44fa7b78cef70efaR58

File: src/components/ChecksVisualization/ChecksVisualization.tsx (hunk 2 above)
Comment (jewbetcha) [reply src/components/ChecksVisualization/ChecksVisualization.tsx:134]: Yes, absolutely 👍

File: src/components/SuccessRateContextProvider.tsx (hunk 3 above)
Comment (jewbetcha) [reply src/components/SuccessRateContextProvider.tsx:103]: Still TODO, hoping for future self to come across it 😛

File: src/components/Thresholds/ThresholdFormSection.tsx (hunk 4 above)
Comment (jewbetcha) [reply src/components/Thresholds/ThresholdFormSection.tsx:50]: Nice, will try

File: src/contexts/SuccessRateContext.ts (hunk 8 above)
Comment (jewbetcha) [reply src/contexts/SuccessRateContext.ts:60]: Good catch 👍

**Pull request merged**